bash-preexec.sh

```diff
@@ -342,8 +371,33 @@ __bp_install_after_session_init() {
     if [[ -n "$sanitized_prompt_command" ]]; then
         PROMPT_COMMAND=${sanitized_prompt_command}$'\n'
     fi;
-    PROMPT_COMMAND+=${__bp_install_string}
+    PROMPT_COMMAND+=${bash_preexec_install_string}
+}
+
+# Remove hooks installed in the DEBUG trap and PROMPT_COMMAND.
+bash_preexec_uninstall() {
+    # Remove __bp_install hook from PROMPT_COMMAND
+    if [[ ${PROMPT_COMMAND-} == *"$bash_preexec_install_string"* ]]; then
+        PROMPT_COMMAND="${PROMPT_COMMAND//$bash_preexec_install_string[;$'\n']}" # Edge case of appending to PROMPT_COMMAND
+        PROMPT_COMMAND="${PROMPT_COMMAND//$bash_preexec_install_string}"
+    fi
+
+    # Remove precmd hook from PROMPT_COMMAND
+    PROMPT_COMMAND=${PROMPT_COMMAND/#$'__bp_precmd_invoke_cmd\n'/$'\n'}
+    PROMPT_COMMAND=${PROMPT_COMMAND%$'\n__bp_interactive_mode'}
+    PROMPT_COMMAND=${PROMPT_COMMAND#$'\n'}
+
+    # Remove preexec hook in the DEBUG trap
+    local q="'" Q="'\''"
+    if [[ $(trap -p DEBUG) == "trap -- '${bash_preexec_trapdebug_string//$q/$Q}' DEBUG" ]]; then
+        if [[ ${__bp_trap_string-} ]]; then
+            eval -- "$__bp_trap_string"
+        else
+            trap - DEBUG
+        fi
+    fi
 }
+declare -ft bash_preexec_uninstall
```

Comment:
any reason this needs to be declared? just good practice?

Reply:
We need it because the effect of `trap - DEBUG` is local to that function call unless the trace attribute is set to the function.

Reply:
01527e5 I have added a code comment.